src/yunohost/backup.py

```diff
@@ -199,257 +300,374 @@ def _clean_tmp_dir(retcode=0):
         else:
             apps_filtered = apps_list
 
-        # Run apps backup scripts
-        tmp_script = '/tmp/backup_' + str(timestamp)
         for app_instance_name in apps_filtered:
-            app_setting_path = '/etc/yunohost/apps/' + app_instance_name
+            self._collect_app_files(app_instance_name)
+
+    def _collect_app_files(self, app_instance_name):
+        app_setting_path = os.path.join('/etc/yunohost/apps/',
+                                        app_instance_name)
+
+        # Check if the app has a backup and restore script
+        app_script = os.path.join(app_setting_path, 'scripts/backup')
+        app_restore_script = os.path.join(app_setting_path, 'scripts/restore')
+        if not os.path.isfile(app_script):
+            logger.warning(m18n.n('unbackup_app', app=app_instance_name))
+            return
+        elif not os.path.isfile(app_restore_script):
+            logger.warning(m18n.n('unrestore_app', app=app_instance_name))
+
+        tmp_app_dir = os.path.join('apps/', app_instance_name)
+        tmp_app_bkp_dir = os.path.join(self.collect_dir, tmp_app_dir, 'backup')
+        logger.info(m18n.n('backup_running_app_script', app=app_instance_name))
+        try:
+            # Prepare backup directory for the app
+            filesystem.mkdir(tmp_app_bkp_dir, 0750, True, uid='admin')
+            settings_dir = os.path.join(tmp_app_dir, 'settings')
+
+            # Copy app backup script in a temporary folder and execute it
+            _, tmp_script = tempfile.mkstemp(prefix='backup_')
+            subprocess.call(['install', '-Dm555', app_script, tmp_script])
+
+            # Prepare env. var. to pass to script
+            app_id, app_instance_nb = _parse_app_instance_name(
+                app_instance_name)
+            env_dict = self._get_env_var()
+            env_dict["YNH_APP_ID"] = app_id
+            env_dict["YNH_APP_INSTANCE_NAME"] = app_instance_name
+            env_dict["YNH_APP_INSTANCE_NUMBER"] = str(app_instance_nb)
+            env_dict["YNH_APP_BACKUP_DIR"] = tmp_app_bkp_dir
```

Comment:
I don't know if that help and is relevant, but it might be cleaner/useful to have a method that does all that job of "preparing the backup environment for an app" (the env variables, but also the mkdir and the "install -Dm555" (whatever that is)), from a given app_instance_name, and returns the dictionary corresponding to env vars.